subprojects/build-cache-packaging/src/main/java/org/gradle/caching/internal/packaging/impl/RelativePathParser.java

```diff
@@ -19,54 +19,67 @@
+    private final int rootLength;
+    private String currentPath;
     private int sizeOfCommonPrefix;
-    private int rootLength;
-    private Deque<String> directoryPaths = new LinkedList<String>();
+
+    public RelativePathParser(String rootPath) {
+        this.directoryPaths.addLast(rootPath.substring(0, rootPath.length() - 1));
+        this.rootLength = rootPath.length();
+        this.currentPath  = rootPath;
+    }
 
     public String getRelativePath() {
-        return currentName.substring(rootLength);
+        return currentPath.substring(rootLength);
     }
 
     public String getName() {
-        return currentName.substring(sizeOfCommonPrefix + 1);
+        return currentPath.substring(sizeOfCommonPrefix + 1);
     }
 
-    public int nextPath(String nextPath, boolean directory) {
-        currentName = directory ? nextPath.substring(0, nextPath.length() - 1): nextPath;
+    public boolean nextPath(String nextPath, boolean directory, Consumer<String> exitDirectory) {
+        currentPath = directory ? nextPath.substring(0, nextPath.length() - 1): nextPath;
         String lastDirPath = directoryPaths.peekLast();
-        sizeOfCommonPrefix = FilePathUtil.sizeOfCommonPrefix(lastDirPath, currentName, 0, '/');
+        sizeOfCommonPrefix = FilePathUtil.sizeOfCommonPrefix(lastDirPath, currentPath, 0, '/');
         int directoriesLeft = determineDirectoriesLeft(lastDirPath, sizeOfCommonPrefix);
         for (int i = 0; i < directoriesLeft; i++) {
             directoryPaths.removeLast();
+            String name = directoryNames.pollLast();
+            if (name == null) {
+                return true;
+            }
+            exitDirectory.accept(name);
         }
-        if (directory && getDepth() > 0) {
-            directoryPaths.addLast(currentName);
+        String currentName = currentPath.substring(sizeOfCommonPrefix + 1);
+        if (directory) {
+            directoryPaths.addLast(currentPath);
+            directoryNames.addLast(currentName);
         }
-        return directoriesLeft;
+        return isRoot();
     }
 
-    private int determineDirectoriesLeft(String lastDirPath, int sizeOfCommonPrefix) {
+    private static int determineDirectoriesLeft(String lastDirPath, int sizeOfCommonPrefix) {
         if (sizeOfCommonPrefix == lastDirPath.length()) {
             return 0;
         }
         int rootDirAdjustment = (sizeOfCommonPrefix == 0) ? 1 : 0;
         return rootDirAdjustment + IS_SLASH.countIn(lastDirPath.substring(sizeOfCommonPrefix));
     }
 
-    public void rootPath(String path) {
-        directoryPaths.addLast(path.substring(0, path.length() - 1));
-        rootLength = path.length();
+    public boolean isRoot() {
+        return directoryNames.isEmpty() && currentPath.length() == rootLength;
     }
 
-    public int getDepth() {
-        return directoryPaths.size();
+    public void exitHierarchy(Consumer<String> handler) {
```

Comment:
Should this be `exitHierarchies()`? Or, to be consistent with the other visitors, `leaveHierarchies()`?

Reply:
I polished the API a little. I'm now using `exit` everywhere, as `left` was ambiguous for me; when I first looked at the code I thought it meant "remaining".